pandas/core/categorical.py

```diff
@@ -1665,16 +1665,26 @@ def fillna(self, value=None, method=None, limit=None):
 
         else:
 
-            if not isna(value) and value not in self.categories:
-                raise ValueError("fill value must be in categories")
+            if isinstance(value, ABCSeries):
+                if not value[~value.isin(self.categories)].isna().all():
+                    raise ValueError("fill value must be in categories")
 
-            mask = values == -1
-            if mask.any():
-                values = values.copy()
-                if isna(value):
-                    values[mask] = -1
-                else:
-                    values[mask] = self.categories.get_loc(value)
+                values_codes = _get_codes_for_values(value, self.categories)
+                indexer = np.where(values_codes != -1)
+                values[indexer] = values_codes[values_codes != -1]
+
+            # Scalar value
+            else:
+                if not isna(value) and value not in self.categories:
+                    raise ValueError("fill value must be in categories")
+
+                mask = values == -1
+                if mask.any():
+                    values = values.copy()
+                    if isna(value):
```

Comment:
you can simplify this to

``values[mask] = self.categories.get_indexer([value])[0]``

Reply:
``get_indexer`` seems to cause problems with ``PeriodIndex``. When I replace the code with your suggestion:
```
In[1]: idx = pd.PeriodIndex(['2011-01', '2011-01', pd.NaT], freq='M')
In[2]: df = pd.DataFrame({'a': pd.Categorical(idx)})

In [3]: df.fillna(value=pd.NaT))
Out[3]:  
pandas._libs.period.IncompatibleFrequency: Input has different freq=None from PeriodIndex(freq=M)
```

Reply:
hmm, this my be a known bug